Firestore/core/src/firebase/firestore/util/to_string.h

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_TO_STRING_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_TO_STRING_H_
+
+#if __OBJC__
+#import <Foundation/Foundation.h>
+#endif  // __OBJC__
+
+#include <algorithm>
+#include <numeric>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+#include "Firestore/core/src/firebase/firestore/util/type_traits.h"
+#include "absl/meta/type_traits.h"
+#include "absl/strings/str_join.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+template <typename T>
+std::string ToString(const T& value);
+
+namespace impl {
+
+// Checks whether the given type `T` defines a member function `ToString`
+
+template <typename T, typename = absl::void_t<>>
+struct has_to_string : std::false_type {};
+
+template <typename T>
+struct has_to_string<T, absl::void_t<decltype(std::declval<T>().ToString())>>
+    : std::true_type {};
+
+template <int I>
+struct Choice : Choice<I + 1> {};
+
+template <>
+struct Choice<5> {};
+
+#if __OBJC__
+
+// Objective-C class
+template <typename T,
+          typename = absl::enable_if_t<is_objective_c_pointer<T>::value>>
+std::string ToStringImpl(T value, Choice<0>) {
+  return MakeString([value description]);
+}
+
+#endif  // __OBJC__
+
+// Has `ToString` member function
+template <typename T, typename = absl::enable_if_t<has_to_string<T>::value>>
+std::string ToStringImpl(const T& value, Choice<1>) {
+  return value.ToString();
+}
+
+// `std::string`
+template <typename T,
+          typename = absl::enable_if_t<std::is_same<std::string, T>::value>>
+std::string ToStringImpl(const T& value, Choice<2>) {
+  return value;
+}
+
+// Associative container
+template <typename T,
+          typename = absl::enable_if_t<is_associative_container<T>::value>>
+std::string ToStringImpl(const T& value, Choice<3>) {
+  std::string contents = absl::StrJoin(
+      value, ", ", [](std::string* out, const typename T::value_type& kv) {
+        out->append(
+            StringFormat("%s: %s", ToString(kv.first), ToString(kv.second)));
```

Comment:
StringFormat is pretty heavyweight for this, and it would be useful to make these implementations not mutually recursive. Consider just using three append calls.